Add `repo license list/view` and `repo gitignore list/view`

BagToad opened the pull request:
This PR resolves #8946 according to the interface described in https://github.com/cli/cli/issues/8946#issuecomment-2217449697

## Examples

### `gh repo license list`

<details><summary>Details</summary>
<p>

![image](https://github.com/user-attachments/assets/d0964109-a8ae-4a41-8779-9f23a3647008)

</p>
</details> 

### `gh repo license view <license>`

<details><summary>Details</summary>
<p>

![image](https://github.com/user-attachments/assets/ce9ad42e-62a3-471b-aa3a-b09cc814ef87)

</p>
</details> 

### `gh repo gitignore list`

<details><summary>Details</summary>
<p>

![image](https://github.com/user-attachments/assets/1cf055d4-1e41-40c3-88c1-e6eda27c5690)

Note: truncated screenshot because the list is long. 

</p>
</details> 

### `gh repo gitignore view <gitignore-template>`

<details><summary>Details</summary>
<p>

![image](https://github.com/user-attachments/assets/ce9f8fa4-0c9d-4024-92f7-529423229e34)

Note: truncated screenshot because the output is long. 

</p>
</details> 

## Notes

- This PR moves and exports existing methods `listLicenseTemplates` and `listGitIgnoreTemplates` found here: https://github.com/cli/cli/blob/d645fd4f0066d96b1e5ef37089192f0d33ba9b84/pkg/cmd/repo/create/http.go#L342-L362 to `queries_repo.go`. I felt this made the most sense since `gh repo create` uses these methods as well.
- I chose to go with table output for the list commands rather than just dumping out a list. Even thought `gitignore list` is a single column, I felt it would make sense to use the same tableprinter anyway.
- ~~You may notice that this PR renames `shared` to `format` in `pkg/cmd/repo/create/create.go` - this is because I had originally missed that there were many queries in `queries_repo.go`; instead of using `queries_repo.go`, I had implemented queries in a package under `shared`, which prompted me to split `shared` into `format` and `queries` to better organize the code. However, I later removed `queries` when I realized about `queries_repo.go`, but decided to leave `format` as a changed package to align more closely with the package structure found in some other examples like https://github.com/cli/cli/tree/d645fd4f0066d96b1e5ef37089192f0d33ba9b84/pkg/cmd/project/shared/format~~ This was reversed. 

## Feedback

I would love constructive feedback on anything, but I am particularly interested in:

- Did I miss any testing opportunities?
- Comments about the mild restructure and migration of listing functions
- ~~Do you agree/disagree with the rename of the `shared` package to format? See Notes section above for more info.~~
This was reversed.

BagToad committed:
First pass at implementing `gh repo license list`

BagToad committed:
Divide shared repo package and add queries tests

BagToad committed:
Remove json output option

Because the API query uses REST, the JSON exporter doesn't work as expected.

BagToad committed:
Cleanup rendering and tests

BagToad committed:
fix output capitalization

BagToad committed:
Fix ListLicenseTemplates doc

BagToad committed:
Update license table headings and tests

BagToad committed:
implement gitignore list

BagToad committed:
Implement gitignore view

BagToad committed:
Implement license view

BagToad committed:
Move API calls to queries_repo.go

BagToad committed:
Refactor tests and add new tests

BagToad committed:
Merge branch 'trunk' into bagtoad/cli-8946-list-licenses-and-gitignore-options

andyfeller commented on pkg/cmd/repo/shared/format/format.go:
```diff
@@ -1,4 +1,4 @@
-package shared
+package format
```
**Unless there is an objective need for the special package namespacing, I think this should be reverted.**

I know you explained this in the PR body, so not rehashing any of that.

Though `gh project` does this, I don't know if it served an objective purpose like specially namespacing types for testing perspective.  Otherwise, it is unclear whether this is for individual aesthetic reasons or has a purpose.

BagToad replied:
Sounds good 👍

andyfeller commented on api/queries_repo.go:
```diff
@@ -207,8 +207,24 @@ type IssueLabel struct {
+	SPDXID         string   `json:"spdx_id"`
+	URL            string   `json:"url"`
+	NodeID         string   `json:"node_id"`
+	HTMLURL        string   `json:"html_url"`
```
I know we do this elsewhere in the code base but frankly the Go conventions around acronyms don't always make sense 🤦 ... `ID` is an abbreviation for identifier, not an acronym

Nothing to change here, just venting as ☝️ isn't really human friendly but it is idiomatic

andyfeller commented on api/queries_repo.go:
```diff
@@ -1391,3 +1407,51 @@ func RepoExists(client *Client, repo ghrepo.Interface) (bool, error) {
+func LicenseTemplate(httpClient *http.Client, hostname string, licenseTemplateName string) (License, error) {
+	var licenseTemplates License
+	client := NewClientFromHTTP(httpClient)
+	err := client.REST(hostname, "GET", fmt.Sprintf("licenses/%v", licenseTemplateName), nil, &licenseTemplates)
```
Minor nit for keeping this in line with [similar logic](https://github.com/search?q=repo%3Acli%2Fcli%20%22client.REST%22&type=code), avoid complex single statements:

```suggestion
	path := fmt.Sprintf("licenses/%s", licenseTemplateName)
	err := client.REST(hostname, "GET", path, nil, &licenseTemplates)
```

andyfeller commented on api/queries_repo.go:
```diff
@@ -1391,3 +1407,51 @@ func RepoExists(client *Client, repo ghrepo.Interface) (bool, error) {
 	}
 }
+
+// ListLicenseTemplates fetches available repository license templates.
+// It uses API v3 because license template isn't supported by GraphQL.
+func ListLicenseTemplates(httpClient *http.Client, hostname string) ([]License, error) {
+	var licenseTemplates []License
```
I caution using the _"template"_ language here because OSS licenses are not choose your own adventure to modify.  It is also not used in official GitHub docs for the endpoints or licensing.

That said, I recognize `gitignore` are templates and are modified by developers, so that is totally cool.

BagToad replied:
Good call - we should change this. What do you think about just `ListLicenses`?

BagToad replied:
Updated to `RepoLicenses` and `RepoLicense`. Also updated the gitignore functions to `RepoGitIgnoreTemplates` and `RepoGitIgnoreTemplate` to be consistent.

andyfeller commented on api/queries_repo.go:
```diff
@@ -1391,3 +1407,51 @@ func RepoExists(client *Client, repo ghrepo.Interface) (bool, error) {
+func GitIgnoreTemplate(httpClient *http.Client, hostname string, gitIgnoreTemplateName string) (GitIgnore, error) {
+	var gitIgnoreTemplateResponse GitIgnore
+	client := NewClientFromHTTP(httpClient)
+	err := client.REST(hostname, "GET", fmt.Sprintf("gitignore/templates/%v", gitIgnoreTemplateName), nil, &gitIgnoreTemplateResponse)
```
```suggestion
	path := fmt.Sprintf("gitignore/templates/%s", gitIgnoreTemplateName)
	err := client.REST(hostname, "GET", path, nil, &gitIgnoreTemplateResponse)
```

andyfeller commented on api/queries_repo.go:
```diff
@@ -1391,3 +1407,51 @@ func RepoExists(client *Client, repo ghrepo.Interface) (bool, error) {
+// It uses API v3 here because gitignore template isn't supported by GraphQL.
+func GitIgnoreTemplate(httpClient *http.Client, hostname string, gitIgnoreTemplateName string) (GitIgnore, error) {
+	var gitIgnoreTemplateResponse GitIgnore
+	client := NewClientFromHTTP(httpClient)
+	err := client.REST(hostname, "GET", fmt.Sprintf("gitignore/templates/%v", gitIgnoreTemplateName), nil, &gitIgnoreTemplateResponse)
+	if err != nil {
+		return GitIgnore{}, err
+	}
```
Normally in cases like this, just return `nil` where possible for the object if an error occurs.

BagToad replied:
I did this because `RepoGitIgnoreTemplate`'s signature is not returning a pointer to a `GitIgnore`. Should I change that? Without any other code changes this is a syntax error.

BagToad replied:
Fixed this 👍

andyfeller commented on pkg/cmd/repo/license/list/list.go:
```diff
@@ -0,0 +1,83 @@
+			}
+			return listRun(opts)
+		},
+	}
```
I know the underlying code is hitting the only GitHub API endpoint available to provide license information.  That said, there are many other types of licenses that are not returned from that endpoint that are covered on https://choosealicense.com/appendix/

**Should there be a `Long` section that explains its limit and where to go for more information?**

BagToad replied:
Done 👍

andyfeller commented on pkg/cmd/repo/license/list/list.go:
```diff
@@ -0,0 +1,83 @@
+		Use:     "list",
+		Short:   "List available repository license templates",
+		Aliases: []string{"ls"},
+		Args:    cmdutil.ExactArgs(0, "gh repo license list takes no arguments"),
```
This error message is somewhat atypical from other places [`cmdutil.ExactArgs` is used](https://github.com/search?q=repo%3Acli%2Fcli+%22cmdutil.ExactArgs%22&type=code) 🤔 

Other commands with no args do something like [`gh auth switch`](https://github.com/cli/cli/blob/dfddb16ba34d5e7b2fb912c7fb6426c5cd2951c5/pkg/cmd/auth/switch/switch.go#L33):

```go
	cmd := &cobra.Command{
		Use:   "switch",
		Args:  cobra.ExactArgs(0),
```

BagToad replied:
Fixed 👍

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -0,0 +1,99 @@
+		HTTPClient: f.HttpClient,
+		Config:     f.Config,
+		License:    "",
+	}
```
Shouldn't need to initialize `License` as its the zero state

```suggestion
	opts := &ViewOptions{
		IO:         f.IOStreams,
		HTTPClient: f.HttpClient,
		Config:     f.Config,
	}
```

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -0,0 +1,99 @@
+		View an available repository license template.
+		
+		%[1]s<license>%[1]s is a license name or SPDX ID.
+		`, "`"),
```
Handful of things:

1. The `Use` statement should be updated to reflect the license **key** or SPDX ID; technically the license name is human readable and not supported

   For example, [`gh issue view`](https://github.com/cli/cli/blob/dfddb16ba34d5e7b2fb912c7fb6426c5cd2951c5/pkg/cmd/issue/view/view.go#L49):

   ```go
   Use:   "view {<number> | <url>}",
   ```

1. Technically the choosealicense.com licenses not returned from the `list` command are valid license keys to use here


   ```shell
   ./bin/gh repo license view cern-ohl-s-2.0 
   ```

   Definitely want to direct users for more information about licenses to choosealicense.com
1. When using `heredoc` package, feel free to indent the text 1 level inward, making it easier to see separate of code

BagToad replied:
Fixed ✨

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -0,0 +1,99 @@
+		
+		%[1]s<license>%[1]s is a license name or SPDX ID.
+		`, "`"),
+		Args: cmdutil.ExactArgs(1, "gh repo licese view only takes a single license argument"),
```
1. There is a typo in `licese`
2. Would align this error message more like other commands without the `gh repo licese view` bit

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -0,0 +1,99 @@
+		defer opts.IO.StopPager()
+	} else {
+		fmt.Fprintf(opts.IO.ErrOut, "failed to start pager: %v\n", err)
+	}
```
Technically works but more idiomatic and consistent if reswizzled like [`gh gist view`](https://github.com/cli/cli/blob/dfddb16ba34d5e7b2fb912c7fb6426c5cd2951c5/pkg/cmd/gist/view/view.go#L128-L131):

```suggestion
	if err := opts.IO.StartPager(); err != nil {
		fmt.Fprintf(opts.IO.ErrOut, "starting pager failed: %v\n", err)
	}
	defer opts.IO.StopPager()
```

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -0,0 +1,99 @@
+	}
+
+	hostname, _ := cfg.Authentication().DefaultHost()
+	license, err := api.LicenseTemplate(client, hostname, opts.License)
+	if err != nil {
+		if strings.Contains(err.Error(), "HTTP 404") {
+			return fmt.Errorf("'%s' is not a valid license template name or SPDX ID. Run `gh repo license list` for options", opts.License)
+		}
+		return err
+	}
```
🤔 Couldn't this check if the error is a HTTP status error and check the code like:

https://github.com/cli/cli/blob/dfddb16ba34d5e7b2fb912c7fb6426c5cd2951c5/pkg/cmd/run/delete/delete.go#L112-L122

BagToad replied:
Thank you! I knew something was wrong with doing this, but wasn't sure how to fix it.

BagToad replied:
Fixed

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -0,0 +1,99 @@
+	"github.com/spf13/cobra"
+)
+
+type ViewOptions struct {
+	IO         *iostreams.IOStreams
+	HTTPClient func() (*http.Client, error)
+	Config     func() (gh.Config, error)
+	License    string
+}
```
**What do you think about a `--web` mode that brings users to the choosealicense.com website?**

For example, https://choosealicense.com/licenses/cern-ohl-s-2.0/

BagToad replied:
I like it. I'll see if this is a low amount of work that we can squeeze into this PR.

BagToad replied:
Done 👍

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -0,0 +1,99 @@
+	return renderLicense(license, opts)
+}
+
+func renderLicense(licenseTemplate api.License, opts *ViewOptions) error {
+	cs := opts.IO.ColorScheme()
+	var out strings.Builder
+	if opts.IO.IsStdoutTTY() {
+		out.WriteString(fmt.Sprintf("\n%s\n", cs.Gray(licenseTemplate.Description)))
+		out.WriteString(fmt.Sprintf("\n%s\n", cs.Grayf("To implement: %s", licenseTemplate.Implementation)))
+		out.WriteString(fmt.Sprintf("\n%s\n\n", cs.Grayf("For more information, see: %s", licenseTemplate.HTMLURL)))
+	}
+	out.WriteString(licenseTemplate.Body)
+	_, err := opts.IO.Out.Write([]byte(out.String()))
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
1. **Should users be able to redirect stdout to a license file to contain just the license body?**

   If so, then we would need to use stderr for informational.
1. **Do we use other colors when communicate actions for users to take?**

   Not sure dim / gray is the right choice to direct the user for taking action.

BagToad replied:
> Should users be able to redirect stdout to a license file to contain just the license body?

I think it already does this.

```shell
❯ ./bin/gh repo license view mit > license.txt

❯ cat license.txt 
MIT License

Copyright (c) [year] [fullname]

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

```

> Do we use other colors when communicate actions for users to take?

Open to suggestions 😁 I just wanted it to be distinct from the actual license.

andyfeller replied:
Ooh, I get it. 🤔 Well done.

_I got to assuming that the dimmed text would be stderr so it would be separate from stdout license content, without clicking this wouldn't print the dimmed text if redirected._

andyfeller replied:
I suppose I was thinking about how we use yellow [here](https://github.com/cli/cli/blob/76ea9396275ac5a79d62f6987303b7e12b147ff9/cmd/gh/main.go#L174-L191) for calling attention to `gh` upgrade, which aligns with the emphasis [Primer color role](https://primer.style/foundations/color/overview#color-roles).

That said, it isn't a deal breaker if a specific color; maybe not dimming it?  I'm open to ideas.

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -0,0 +1,99 @@
+			}
+			return viewRun(opts)
+		},
+	}
```
**Should we have an `Examples` section?**

If users should be able to generate a license file from this command involving redirecting stdout, then we could demonstrate the various ways to invoke it. 🤷

BagToad replied:
Added example section with an example for redirecting stdout

andyfeller commented on pkg/cmd/repo/license/view/view_test.go:
```diff
@@ -0,0 +1,263 @@
+		tty      bool
+	}{
+		{
+			name:    "No license name or SPDX ID provided",
```
```suggestion
			name:    "No license key or SPDX ID provided",
```

andyfeller commented on pkg/cmd/repo/license/view/view_test.go:
```diff
@@ -0,0 +1,263 @@
+			},
+		},
+		{
+			name:    "Happy path single license name",
```
```suggestion
			name:    "Happy path single license key",
```

andyfeller commented on pkg/cmd/repo/license/view/view_test.go:
```diff
@@ -0,0 +1,263 @@
+				License: "mit",
+			},
+		},
+		{
+			name:    "Happy path too many license names",
+			args:    []string{"mit", "apache-2.0"},
+			wantErr: true,
+			wantOpts: &ViewOptions{
+				License: "",
+			},
+		},
```
```suggestion
		{
			name:    "Happy path too many license keys",
			args:    []string{"mit", "apache-2.0"},
			wantErr: true,
			wantOpts: &ViewOptions{
				License: "",
			},
		},
```

andyfeller commented on pkg/cmd/repo/license/view/view_test.go:
```diff
@@ -0,0 +1,263 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewCmdView(t *testing.T) {
```
**How should this test SPDX ID?**

The examples below are primarily license keys.  Granted most SPDX IDs are just capitalized versions of the license keys but I haven't done an holistic confirmation 🤷

BagToad replied:
I'm not sure if it makes sense to test for both the key and the SPDX ID since we would just hardcode the exact same API response in both cases. 

I don't think I could make one of these tests fail without the other test also failing, unless I screw up the HTTP stubs in the test.

What are your thoughts?

BagToad replied:
Also, to hopefully illustrate my point more, the endpoint appears to be completely case insensitive. All these examples work:

```
gh repo license view MIt
gh repo license view MiT
gh repo license view miT
```

andyfeller replied:
You make a fair point as I wrestled over the same question knowing that license keys and SPDX IDs should be the same, so its fair to punt because it isn't significant.  Making sure to ask the question while we're here 👍

andyfeller commented on pkg/cmd/repo/gitignore/list/list.go:
```diff
@@ -0,0 +1,82 @@
+
+	cmd := &cobra.Command{
+		Use:     "list",
+		Short:   "List available repository gitignore templates",
```
🤔 how does something like this sound instead:

```suggestion
		Short:   "List gitignore templates used in creating repositories",
```

andyfeller commented on pkg/cmd/repo/gitignore/view/view.go:
```diff
@@ -0,0 +1,101 @@
+		Template:   "",
+	}
+
+	cmd := &cobra.Command{
+		Use:   "view <template>",
+		Short: "View an available repository gitignore template",
+		Long: heredoc.Docf(`
+		View an available repository %[1]s.gitignore%[1]s template.
+		
+		%[1]s<template>%[1]s is a case-sensitive %[1]s.gitignore%[1]s template name.
+		`, "`"),
```
Aside from indenting the `Long` description, I don't know if positional arguments are formatted this way in other commands 🤔

At the least, we could have a `Examples` section with common gitignore language templates to demonstrate. 🤷

BagToad replied:
>  indenting the Long description

Fixed 👍 

> I don't know if positional arguments are formatted this way in other commands 🤔

Seems they are - can we leave it as-is since there is prescident for this? 

https://github.com/cli/cli/blob/76ea9396275ac5a79d62f6987303b7e12b147ff9/pkg/cmd/codespace/ssh.go#L64

https://github.com/cli/cli/blob/76ea9396275ac5a79d62f6987303b7e12b147ff9/pkg/cmd/run/download/download.go#L53

https://github.com/cli/cli/blob/76ea9396275ac5a79d62f6987303b7e12b147ff9/pkg/cmd/run/rerun/rerun.go#L51

> At the least, we could have a Examples section with common gitignore language templates to demonstrate. 

Yep added examples to `gitignore view` and `license view` as well 👍

BagToad replied:
Nevermind on the second point - I found a different way to word this 👍

andyfeller commented on pkg/cmd/repo/gitignore/view/view.go:
```diff
@@ -0,0 +1,101 @@
+		
+		%[1]s<template>%[1]s is a case-sensitive %[1]s.gitignore%[1]s template name.
+		`, "`"),
+		Args: cmdutil.ExactArgs(1, "gh repo gitignore view only takes a single template argument"),
```
Again on bringing this into alignment with other commands language

BagToad replied:
Fixed 👍

andyfeller commented on pkg/cmd/repo/gitignore/list/list.go:
```diff
@@ -0,0 +1,82 @@
+		return err
+	}
+
+	if err := opts.IO.StartPager(); err == nil {
+		defer opts.IO.StopPager()
+	} else {
+		fmt.Fprintf(opts.IO.ErrOut, "failed to start pager: %v\n", err)
+	}
```
```suggestion
	if err := opts.IO.StartPager(); err != nil {
		fmt.Fprintf(opts.IO.ErrOut, "starting pager failed: %v\n", err)
	}
	defer opts.IO.StopPager()
```

BagToad replied:
FWIW I think I copied this from somewhere 😆 Totally down to improve it though.

andyfeller commented on pkg/cmd/repo/gitignore/view/view.go:
```diff
@@ -0,0 +1,101 @@
+		return err
+	}
+
+	if err := opts.IO.StartPager(); err == nil {
+		defer opts.IO.StopPager()
+	} else {
+		fmt.Fprintf(opts.IO.ErrOut, "failed to start pager: %v\n", err)
+	}
```
```suggestion
	if err := opts.IO.StartPager(); err != nil {
		fmt.Fprintf(opts.IO.ErrOut, "starting pager failed: %v\n", err)
	}
	defer opts.IO.StopPager()
```

andyfeller commented on pkg/cmd/repo/license/list/list.go:
```diff
@@ -0,0 +1,83 @@
+		return err
+	}
+
+	if err := opts.IO.StartPager(); err == nil {
+		defer opts.IO.StopPager()
+	} else {
+		fmt.Fprintf(opts.IO.ErrOut, "failed to start pager: %v\n", err)
+	}
```
```suggestion
	if err := opts.IO.StartPager(); err != nil {
		fmt.Fprintf(opts.IO.ErrOut, "starting pager failed: %v\n", err)
	}
	defer opts.IO.StopPager()
```

andyfeller commented on pkg/cmd/repo/license/list/list.go:
```diff
@@ -0,0 +1,83 @@
+		return err
+	}
+
+	if len(licenseTemplates) == 0 {
+		return cmdutil.NewNoResultsError("No repository license templates found")
+	}
```
Could you elaborate on where this might occur?

BagToad replied:
Probably never - I suppose you'd need a GitHub host messed up in just the right way for the API to return nothing, but still be a `2xx` HTTP status? 

What do you think? Should we delete this? I prefer to be resilient against an API doing weird things even if you or I don't think this could realistically happen.

andyfeller replied:
I'm okay leaving and seeing what happens; was mostly curious if it occurred during development or not. 🤷

andyfeller commented on pkg/cmd/repo/license/list/list.go:
```diff
@@ -0,0 +1,83 @@
+	return renderLicenseTemplatesTable(licenseTemplates, opts)
+}
+
+func renderLicenseTemplatesTable(licenseTemplates []api.License, opts *ListOptions) error {
+	t := tableprinter.New(opts.IO, tableprinter.WithHeader("LICENSE KEY", "LICENSE NAME"))
+	for _, l := range licenseTemplates {
+		t.AddField(l.Key)
+		t.AddField(l.Name)
+		t.EndRow()
+	}
+
+	return t.Render()
+}
```
**Should we include the SPDX ID in the listing too?**

Since it is an alternative positional argument form...

BagToad replied:
Added

andyfeller reviewed (commented):
I'm sure `gh` users will be happy for this information readily at hand! ✨

No major concerns but a range of various minor or middling nits. 🙇

BagToad committed:
Apply suggestions from code review

Co-authored-by: Andy Feller <andyfeller@github.com>

BagToad committed:
Reverse repo/shared package name change

BagToad committed:
Update function names

Licenses are no longer referred to as templates. Prefix the new license and gitignore functions with "Repo" as a more descriptive name.

BagToad committed:
Add `Long` for license list indicating limitations

BagToad committed:
Fix ExactArgs invocation

BagToad committed:
Add SPDX ID to license list output

BagToad committed:
Update help and fix heredoc indentation

BagToad committed:
Fix license view help doc, add LICENSE.md example

BagToad committed:
implement `--web` flag for license view

BagToad committed:
Refactor http error handling

BagToad committed:
Update help for license view

Updated `Use` to indicate that args can be an SPDX ID or license key.
Improved the examples docs to show the use of both.

BagToad commented:
@andyfeller This is ready for another pass please 🙏 I've implemented most of your suggestions. There are a few unresolved comments that need more discussion.

BagToad committed:
Merge branch 'trunk' into bagtoad/cli-8946-list-licenses-and-gitignore-options

BagToad committed:
Add examples for creating `.gitignore` files

BagToad committed:
Merge branch 'bagtoad/cli-8946-list-licenses-and-gitignore-options' of https://github.com/cli/cli into bagtoad/cli-8946-list-licenses-and-gitignore-options

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -64,21 +89,31 @@ func viewRun(opts *ViewOptions) error {
 		return err
 	}
 
+	if opts.Web {
+		url := fmt.Sprintf("https://choosealicense.com/licenses/%s", license.Key)
+		if opts.IO.IsStdoutTTY() {
+			fmt.Fprintf(opts.IO.Out, "Opening %s in your browser.\n", text.DisplayURL(url))
+		}
+		return opts.Browser.Browse(url)
+	}
+
```
![bongo-cat](https://github.com/user-attachments/assets/31319f66-5abe-47ae-b282-4803dcd88c2c)

andyfeller commented on pkg/cmd/repo/license/list/list.go:
```diff
@@ -26,10 +27,15 @@ func NewCmdList(f *cmdutil.Factory, runF func(*ListOptions) error) *cobra.Comman
+			List available repository license templates.
+			
+			Only the most commonly used licenses templates are returned. For even more licenses, visit <https://choosealicense.com/appendix>
+		`),
```
Aside from removing the _"template"_ language, it seemed like the short and long descriptions were a little confusing if we came back and say _"Well, only the most common used ones ..."_

```suggestion
		Use:   "list",
		Short: "List common repository licenses",
		Long: heredoc.Doc(`
			List common repository licenses.
			
			For even more licenses, visit <https://choosealicense.com/appendix>
		`),
```

andyfeller commented on pkg/cmd/repo/license/license.go:
```diff
@@ -0,0 +1,20 @@
+func NewCmdLicense(f *cmdutil.Factory) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "license <command>",
+		Short: "View available repository license templates",
```
Figuring out how to explain a commandset is a bit rough and none of the other examples really line up.  Since we're only making license information available, I thought maybe the following would be short and sweet:

```suggestion
		Short: "Explore repository licenses",
```

andyfeller commented on pkg/cmd/repo/license/list/list.go:
```diff
@@ -0,0 +1,89 @@
+	defer opts.IO.StopPager()
+
+	hostname, _ := cfg.Authentication().DefaultHost()
+	licenseTemplates, err := api.RepoLicenses(client, hostname)
```
```suggestion
	licenses, err := api.RepoLicenses(client, hostname)
```

andyfeller commented on pkg/cmd/repo/license/list/list.go:
```diff
@@ -0,0 +1,89 @@
+		return err
+	}
+
+	if len(licenseTemplates) == 0 {
+		return cmdutil.NewNoResultsError("No repository license templates found")
+	}
```
```suggestion
	if len(licenses) == 0 {
		return cmdutil.NewNoResultsError("No repository licenses found")
	}
```

andyfeller commented on pkg/cmd/repo/license/list/list.go:
```diff
@@ -0,0 +1,89 @@
+		return cmdutil.NewNoResultsError("No repository license templates found")
+	}
+
+	return renderLicenseTemplatesTable(licenseTemplates, opts)
+}
+
+func renderLicenseTemplatesTable(licenseTemplates []api.License, opts *ListOptions) error {
+	t := tableprinter.New(opts.IO, tableprinter.WithHeader("LICENSE KEY", "SPDX ID", "LICENSE NAME"))
+	for _, l := range licenseTemplates {
+		t.AddField(l.Key)
+		t.AddField(l.SPDXID)
+		t.AddField(l.Name)
+		t.EndRow()
+	}
```
```suggestion
	return renderLicensesTable(licenses, opts)
}

func renderLicensesTable(licenses []api.License, opts *ListOptions) error {
	t := tableprinter.New(opts.IO, tableprinter.WithHeader("LICENSE KEY", "SPDX ID", "LICENSE NAME"))
	for _, l := range licenses {
		t.AddField(l.Key)
		t.AddField(l.SPDXID)
		t.AddField(l.Name)
		t.EndRow()
	}
```

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -0,0 +1,134 @@
+
+	cmd := &cobra.Command{
+		Use:   "view {<license-key> | <SPDX-ID>}",
+		Short: "View an available repository license template",
+		Long: heredoc.Docf(`
+			View an available repository license template.
+			
+			Argument can be a case-insensitive license key or SPDX ID.
+
+			Run %[1]sgh repo license list%[1]s to see available commonly used licenses. For even more licenses, visit <https://choosealicense.com/appendix>.
+		`, "`"),
```
```suggestion
		Short: "View a specific repository license",
		Long: heredoc.Docf(`
			View a specific repository license by license key or SPDX ID.

			Run %[1]sgh repo license list%[1]s to see available commonly used licenses. For even more licenses, visit <https://choosealicense.com/appendix>.
		`, "`"),
```

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -0,0 +1,134 @@
+		var httpErr api.HTTPError
+		if errors.As(err, &httpErr) {
+			if httpErr.StatusCode == 404 {
+				return fmt.Errorf("'%s' is not a valid license template name or SPDX ID.\n\nRun `gh repo license list` to see available commonly used licenses. For even more licenses, visit %s", opts.License, text.DisplayURL("https://choosealicense.com/appendix"))
```
```suggestion
				return fmt.Errorf("'%s' is not a valid license name or SPDX ID.\n\nRun `gh repo license list` to see available commonly used licenses. For even more licenses, visit %s", opts.License, text.DisplayURL("https://choosealicense.com/appendix"))
```

andyfeller commented on pkg/cmd/repo/license/view/view.go:
```diff
@@ -0,0 +1,134 @@
+	return renderLicense(license, opts)
+}
+
+func renderLicense(licenseTemplate api.License, opts *ViewOptions) error {
+	cs := opts.IO.ColorScheme()
+	var out strings.Builder
+	if opts.IO.IsStdoutTTY() {
+		out.WriteString(fmt.Sprintf("\n%s\n", cs.Gray(licenseTemplate.Description)))
+		out.WriteString(fmt.Sprintf("\n%s\n", cs.Grayf("To implement: %s", licenseTemplate.Implementation)))
+		out.WriteString(fmt.Sprintf("\n%s\n\n", cs.Grayf("For more information, see: %s", licenseTemplate.HTMLURL)))
+	}
+	out.WriteString(licenseTemplate.Body)
+	_, err := opts.IO.Out.Write([]byte(out.String()))
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
```suggestion
func renderLicense(license api.License, opts *ViewOptions) error {
	cs := opts.IO.ColorScheme()
	var out strings.Builder
	if opts.IO.IsStdoutTTY() {
		out.WriteString(fmt.Sprintf("\n%s\n", cs.Gray(license.Description)))
		out.WriteString(fmt.Sprintf("\n%s\n", cs.Grayf("To implement: %s", license.Implementation)))
		out.WriteString(fmt.Sprintf("\n%s\n\n", cs.Grayf("For more information, see: %s", license.HTMLURL)))
	}
	out.WriteString(license.Body)
	_, err := opts.IO.Out.Write([]byte(out.String()))
	if err != nil {
		return err
	}
	return nil
}
```

andyfeller reviewed (commented):
Final round of minor nits, primarily around language.  Happy to knock this out with you today if you like.

BagToad committed:
Apply suggestions from code review

Co-authored-by: Andy Feller <andyfeller@github.com>

BagToad committed:
Fix license list and view tests

BagToad committed:
Return nil instead of empty objects when err

andyfeller reviewed (approved):
✨ 👨‍🍳 💋

BagToad committed:
Fix api tests after function signature changes

yermulnik commented:
User feedback FWIW: reading [v2.59.0](https://github.com/cli/cli/releases/tag/v2.59.0) `What's Changed` section I noticed nifty `repo gitignore list/view` item, which I thought was about repo's `.gitignore` file(s). Though apparently looking into this PR I regretfully found this feature is about gitignore **templates** 🤷🏻 Would be great to eliminate this confusion by replacing `gitignore` sub-command with e.g. `gitignore-templates` or similar. Thanks.
Same probably applies to `license` sub-command too.

yermulnik commented:
> Would be great to eliminate this confusion by replacing `gitignore` sub-command with e.g. `gitignore-templates` or similar. Thanks.
> Same probably applies to `license` sub-command too.

Or maybe even by adding `template` sub-command with nested `license` and `gitignore` sub-commands 🤔

andyfeller commented:
> User feedback FWIW: reading [v2.59.0](https://github.com/cli/cli/releases/tag/v2.59.0) `What's Changed` section I noticed nifty `repo gitignore list/view` item, which I thought was about repo's `.gitignore` file(s). Though apparently looking into this PR I regretfully found this feature is about gitignore **templates** 🤷🏻 Would be great to eliminate this confusion by replacing `gitignore` sub-command with e.g. `gitignore-templates` or similar. Thanks. Same probably applies to `license` sub-command too.

@yermulnik : I'd love to follow up within the original issue or a new issue about your ideas on what `gh` might do with existing `.gitignore` files within the repository.

Normally, I'd say `gh` is focused on helping people focus on the primitives and information across GitHub while leaving `git` for users working with the contents within the repository.  The original issue here was helping developers creating repositories understand what various licenses and `.gitignore` templates entailed and generating the initial content.

yermulnik commented:
> @yermulnik : I'd love to follow up within the original issue or a new issue about your ideas on what `gh` might do with existing `.gitignore` files within the repository.

Apologies if I wasn't clear enough: this is not about handling `.gitignore` files, but rather this is about misleading sub-command naming.

> […] various licenses and `.gitignore` templates […]

This was the gist of my comment: the word "**templates**", that you used to make thought clear, is missing from the added sub-commands and this adds a bit of confusion I guess.
